### PR TITLE
Add service to provide concurrent safe IPAM

### PR DESCRIPTION
**Pull request by tuommaki**
With Node Pools there is a need for performing network allocations from
multiple controllers concurrently. In order to make it correctly, start by
creating an implementation that performs inventory on existing networks and
persisting of allocated network within a lock so that there is no race
condition between callers.

Later with VOO the in-process mutex must be changed to be a distributed lock in
order to maintain correct behavior.

Towards Node Pools: https://github.com/giantswarm/giantswarm/issues/5833

### CODE REVIEW

File: service/ipam/spec.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	// used for getting currently reserved networks and persisting allocated
+	// network. These are called within an single lock so given that callbacks
+	// work correctly, this is concurrent safe within single process e.g. when
+	// logic between multiple controllers are allocating networks from the same
+	// pool.
+	//
+	// NOTE: This is NOT concurrent safe between distinct processes.
+	//
+	// This returns either allocated network or an error.
+	Allocate(ctx context.Context, fullRange net.IPNet, netSize net.IPMask, callbacks AllocationCallbacks) (net.IPNet, error)
```
Comment (xh3b4sd) [inline service/ipam/spec.go]: I do not follow that callback thing. Why do we need that? To me it looks like a way to configure IPAM to break. What is the idea behind that?

Comment (tuommaki) [reply service/ipam/spec.go]: For different clusters there are different AWS clients which affect the way how existing networks are gathered. I don't want to mix handling of those inside the `NetworkAllocator` - it's not really its business.

Similarly persisting allocated network means different thing for old code and new code (whether it's written to `AWSConfig` or `Cluster` CR). This also doesn't belong to `NetworkAllocator`.

In order to have correctness, both gathering existing networks and persisting allocated new subnet must be performed inside a single lock. That's why those implementations are passed as callbacks to this service.

Comment (xh3b4sd) [reply service/ipam/spec.go]: With you.

**Comment by JosephSalisbury** [comment]
Linking https://github.com/giantswarm/giantswarm/issues/5275, as the VOO story for ensuring operators don't clobber each other

**Pull request merged**